src/System.Runtime.Extensions/src/System/Environment.CoreCLR.cs

```diff
@@ -0,0 +1,406 @@
+    {
+        public static int ExitCode { get { return EnvironmentAugments.ExitCode; } set { EnvironmentAugments.ExitCode = value; } }
+
+        private static string ExpandEnvironmentVariablesCore(string name)
+        {
+            int currentSize = 100;
+            StringBuilder result = StringBuilderCache.Acquire(currentSize); // A somewhat reasonable default size
+
+            result.Length = 0;
+            int size = Interop.mincore.ExpandEnvironmentStringsW(name, result, currentSize);
+            if (size == 0)
+            {
+                StringBuilderCache.Release(result);
+                Marshal.ThrowExceptionForHR(Marshal.GetHRForLastWin32Error());
+            }
+
+            while (size > currentSize)
+            {
+                currentSize = size;
+                result.Capacity = currentSize;
+                result.Length = 0;
+
+                size = Interop.mincore.ExpandEnvironmentStringsW(name, result, currentSize);
+                if (size == 0)
+                {
+                    StringBuilderCache.Release(result);
+                    Marshal.ThrowExceptionForHR(Marshal.GetHRForLastWin32Error());
+                }
+            }
+
+            return StringBuilderCache.GetStringAndRelease(result);
+        }
+
+        private static string GetEnvironmentVariableCore(string variable, EnvironmentVariableTarget target)
+        {
+            if (target == EnvironmentVariableTarget.Process)
+            {
+                StringBuilder sb = StringBuilderCache.Acquire(128); // a somewhat reasonable default size
+                int requiredSize = Interop.mincore.GetEnvironmentVariableW(variable, sb, sb.Capacity);
+                if (requiredSize == 0 && Marshal.GetLastWin32Error() == Interop.mincore.Errors.ERROR_ENVVAR_NOT_FOUND)
+                {
+                    StringBuilderCache.Release(sb);
+                    return null;
+                }
+
+                while (requiredSize > sb.Capacity)
+                {
+                    sb.Capacity = requiredSize;
+                    sb.Length = 0;
+                    requiredSize = Interop.mincore.GetEnvironmentVariableW(variable, sb, sb.Capacity);
+                }
+
+                return StringBuilderCache.GetStringAndRelease(sb);
+            }
+            else if (target == EnvironmentVariableTarget.Machine)
+            {
+                // TODO #8533: Uncomment/fix when registry APIs available
+                //using (RegistryKey environmentKey = Registry.LocalMachine.OpenSubKey(@"System\CurrentControlSet\Control\Session Manager\Environment", false))
+                //{
+                //    return environmentKey?.GetValue(variable) as string ?? null;
+                //}
+                return null;
+            }
+            else
+            {
+                Debug.Assert(target == EnvironmentVariableTarget.User);
+
+                // TODO #8533: Uncomment/fix when registry APIs available
```

Comment:
I'm not sure.  If we pull both this and the registry APIs into System.Runtime, then it'd be available, but otherwise it'd be a bad cycle.  We'll probably just need to P/Invoke directly; potentially we could do some small refactoring on Microsoft.Win32.Registry to make the relevant parts easily sharable.